<launch>
  <include file="$(find autorally_core)/launch/hardware.machine" />

  <!-- TODO: remove unused params! -->
  <!-- Debug launch prefix: xterm -e gdb  -->

  <node
    pkg = "autorally_control" 
    type = "path_integral_omni_wheel_robot" 
    name = "mppi_controller"
    output = "screen"
    launch-prefix="" 
    >
    
    <!--Pose topic configuration parameters -->
    <param name="pose_estimate" value="/ground_truth/state" />
    <param name="path_topic" value="/mppi_controller/nominal_path" />

    <!--Debug/Sim mode parameter -->
    <param name="debug_mode" value="false" />

    <!--Setup parameters-->
    <param name="num_timesteps" value="100" />
    <param name="x_pos" value="0" />
    <param name="y_pos" value="0" />
    <param name="heading" value="0" />
    <param name="wheel_speed_exploration_variance" value="0.5" />

    <!-- How many controls are executed between optimization loops -->
    <param name="optimization_stride" value="1" />
    <param name="use_feedback_gains" value="true" />

    <!-- Control hyper-parameters-->
    <param name="norm_exp_kernel_gamma" value=".5" />
    <param name="num_optimization_iters" value="1" />

    <!-- Control parameters -->

    <!-- Control Constraints -->
    <!-- TODO: this is really the max abs per-wheel control, rename -->
    <param name="max_abs_wheel_force" value="1.1" />
    <param name="controls_frequency" value="60" />

    <!-- Cost Parameters -->
    <param name="l1_cost" value="false" />
    <param name="desired_speed" value="2.0" />
    <!--<param name="speed_coefficient" value="4.25" />-->
    <param name="speed_coefficient" value="4.5" />
    <!--<param name="track_coefficient" value="200.0" />-->
    <param name="track_coefficient" value="200.0" />
    <param name="track_slop" value="0.0" />
    <param name="crash_coeff" value="10000.0" />
    <param name="boundary_threshold" value="0.65" />
    <param name="discount" value = "0.1" />
    <param name="map_path" value="$(env AR_MPPI_PARAMS_PATH)/maps/gazebo_costmap_05_22_2016.npz" />

    <!-- Tube-MPC Parameters -->
    <!-- TODO: replace these two bools with an enum somehow? -->
    <param name="use_only_actual_state_controller" value="true" />
    <param name="use_only_predicted_state_controller" value="false" />

  </node>

  <!-- TODO: uncomment this section
  <node
    pkg = "autorally_control" 
    type = "path_integral_monitor" 
    name = "mppi_monitor"
    output = "screen">
  </node>

  <node
    pkg = "autorally_control" 
    type = "lap_stats.py" 
    name = "stat_tracker"
    output = "screen">
    <param name="controller_type" value="mppi_controller" />
    <param name="tag" value="Omni Wheel Robot - Path Integral"/> 
  </node>
  -->

</launch>
